File: regions/system/library/go/retry/circuit_breaker.go
package retry

import (
	"fmt"
	"sync"
	"time"
)

// CircuitBreakerState はサーキットブレーカーの状態。
type CircuitBreakerState int

const (
	// StateClosed は正常状態。
	StateClosed CircuitBreakerState = iota
	// StateOpen はオープン状態（リクエストを拒否）。
	StateOpen
	// StateHalfOpen はハーフオープン状態（テストリクエストを許可）。
	StateHalfOpen
)

func (s CircuitBreakerState) String() string {
	switch s {
	case StateClosed:
		return "Closed"
	case StateOpen:
		return "Open"
	case StateHalfOpen:
		return "HalfOpen"
	default:
		return "Unknown"
	}
}

// CircuitBreakerConfig はサーキットブレーカーの設定。
type CircuitBreakerConfig struct {
	// FailureThreshold はオープンに遷移する失敗回数閾値。
	FailureThreshold int
	// SuccessThreshold はハーフオープンからクローズに遷移する成功回数閾値。
	SuccessThreshold int
	// Timeout はオープンからハーフオープンに遷移するまでの待機時間。
	Timeout time.Duration
}

// DefaultCircuitBreakerConfig はデフォルトのサーキットブレーカー設定を返す。
func DefaultCircuitBreakerConfig() *CircuitBreakerConfig {
	return &CircuitBreakerConfig{
		FailureThreshold: 5,
		SuccessThreshold: 2,
		Timeout:          30 * time.Second,
	}
}

// CircuitBreakerOpenError はサーキットブレーカーがオープン状態のときのエラー。
type CircuitBreakerOpenError struct{}

func (e *CircuitBreakerOpenError) Error() string {
	return fmt.Sprintf("サーキットブレーカーがオープン状態です")
}

// CircuitBreaker はサーキットブレーカーパターンの実装。
type CircuitBreaker struct {
	config       *CircuitBreakerConfig
	mu           sync.Mutex
	state        CircuitBreakerState
	failureCount int
	successCount int
	openedAt     time.Time
}

// NewCircuitBreaker は新しいサーキットブレーカーを生成する。
func NewCircuitBreaker(config *CircuitBreakerConfig) *CircuitBreaker {
	return &CircuitBreaker{
		config: config,
		state:  StateClosed,
	}
}

// IsOpen はサーキットブレーカーがオープン状態かどうかを返す。
// オープン状態でタイムアウトを過ぎた場合、ハーフオープンに遷移して false を返す。
func (cb *CircuitBreaker) IsOpen() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.state == StateOpen {
		if time.Since(cb.openedAt) >= cb.config.Timeout {
			cb.state = StateHalfOpen
			cb.successCount = 0
			return false
		}
		return true
	}
	return false
}

// RecordSuccess は成功を記録する。
func (cb *CircuitBreaker) RecordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case StateHalfOpen:
		cb.successCount++
		if cb.successCount >= cb.config.SuccessThreshold {
			cb.state = StateClosed
			cb.failureCount = 0
			cb.successCount = 0
		}
	case StateClosed:
		cb.failureCount = 0
	}
}

// RecordFailure は失敗を記録する。
func (cb *CircuitBreaker) RecordFailure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	cb.failureCount++
	if cb.failureCount >= cb.config.FailureThreshold {
		cb.state = StateOpen
		cb.openedAt = time.Now()
		cb.failureCount = 0
	}
}

// GetState は現在の状態を返す。
func (cb *CircuitBreaker) GetState() CircuitBreakerState {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	return cb.state
}

File: regions/system/library/dart/retry/lib/src/circuit_breaker.dart
enum CircuitBreakerState { closed, open, halfOpen }

class CircuitBreakerConfig {
  final int failureThreshold;
  final int successThreshold;
  final int timeoutMs;

  const CircuitBreakerConfig({
    this.failureThreshold = 5,
    this.successThreshold = 2,
    this.timeoutMs = 30000,
  });
}

class CircuitBreaker {
  final CircuitBreakerConfig config;
  CircuitBreakerState _state = CircuitBreakerState.closed;
  int _failureCount = 0;
  int _successCount = 0;
  int _openedAt = 0;

  CircuitBreaker({CircuitBreakerConfig? config})
      : config = config ?? const CircuitBreakerConfig();

  CircuitBreakerState get state {
    _checkTimeout();
    return _state;
  }

  bool get isOpen {
    _checkTimeout();
    return _state == CircuitBreakerState.open;
  }

  void recordSuccess() {
    _checkTimeout();
    if (_state == CircuitBreakerState.halfOpen) {
      _successCount++;
      if (_successCount >= config.successThreshold) {
        _state = CircuitBreakerState.closed;
        _failureCount = 0;
        _successCount = 0;
        _openedAt = 0;
      }
    } else if (_state == CircuitBreakerState.closed) {
      _failureCount = 0;
    }
  }

  void recordFailure() {
    _checkTimeout();
    _failureCount++;
    if (_failureCount >= config.failureThreshold) {
      _state = CircuitBreakerState.open;
      _openedAt = DateTime.now().millisecondsSinceEpoch;
      _failureCount = 0;
    }
  }

  void _checkTimeout() {
    if (_state == CircuitBreakerState.open && _openedAt > 0) {
      if (DateTime.now().millisecondsSinceEpoch - _openedAt >=
          config.timeoutMs) {
        _state = CircuitBreakerState.halfOpen;
        _successCount = 0;
      }
    }
  }
}
